### PR TITLE
feat(mysql): handle prepared statement ID reuse edge cases

### PR DESCRIPTION
**Pull request by codeMaestro78**
## Summary

This PR handles MySQL prepared statement ID reuse edge cases that were documented as TODO in `match.go`. When clients prepare, close, and re-prepare the same query, the statement IDs change but the current implementation loses track of this lifecycle, causing mock matching failures.

## Changes Made

- **Add `PreparedStmtHistory` struct** (`wire/util.go`) - Tracks full lifecycle of prepared statements including prepare cycles and closure states
- **Extend `prepEntry` struct** (`replayer/match.go`) - Added `prepareOrder` and `wasClosed` fields to track statement lifecycle in recorded mocks
- **Add `markClosedEntries()` function** - Scans CLOSE commands and marks corresponding entries as closed
- **Add `lookupRecordedQueryByContent()` function** - Enables lookup by query text and prepare order for ID reuse scenarios
- **Update `buildRecordedPrepIndex()`** - Now tracks prepare order per query per connection
- **Record PREP/CLOSE in history** (`replayer/query.go`, `replayer/match.go`) - Runtime tracking of statement lifecycle
- **Add comprehensive unit tests** (`replayer/match_test.go`) - 19 tests covering all edge cases

## Links & References

**Closes:** NA (enhancement addressing existing TODO in codebase)

### 🔗 Related PRs
- NA

### 🐞 Related Issues
- NA (Addresses TODO documented in `pkg/agent/proxy/integrations/mysql/replayer/match.go` lines 1020-1038)

### 📄 Related Documents
- NA

## What type of PR is this? (check all applicable)
- [ ] 📦 Chore
- [x] 🍕 Feature
- [x] 🐞 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [x] ✅ Test
- [ ] 🔁 CI
- [ ] ⏩ Revert

## Added e2e test pipeline?
- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed

## Added comments for hard-to-understand areas?
- [x] 👍 yes

## Added to documentation?
- [ ] 📜 README.md
- [ ] 📓 Wiki
- [x] 🙅 no documentation needed

## Are there any sample code or steps to test the changes?
- [x] 👍 yes, mentioned below

### Testing Steps:
```bash
# Run the unit tests
go test ./pkg/agent/proxy/integrations/mysql/replayer/... -v

# Build verification
go build ./...

# Run all MySQL tests
go test ./pkg/agent/proxy/integrations/mysql/...
```
# Edge Cases Now Handled:
1. **Prepare→Close→Prepare→Close→Prepare→Execute** (same query, different IDs each time)
2. **Multiple prepares without closing** (same query gets multiple active IDs)

# Self Review Done?
 - [x] ✅ yes
 - [ ] ❌ no, because I need help

# Any relevant screenshots, recordings or logs?
Test Output:
Screenshot 2025-12-05 162449

ok go.keploy.io/server/v3/pkg/agent/proxy/integrations/mysql/replayer

# Additional checklist:
- [x] Have you read the [Contributing Guidelines on issues](vscode-file://vscode-app/c:/Users/Devarshi/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html)?
- [x] Have you followed the [PR Semantics guide](vscode-file://vscode-app/c:/Users/Devarshi/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) for naming this PR?
- [x] Have you followed the [Branch Semantics guide](vscode-file://vscode-app/c:/Users/Devarshi/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) for naming your branch?

### CODE REVIEW

**Comment by codeMaestro78** [comment]
I have read the CLA Document and I hereby sign the CLA

File: pkg/agent/proxy/integrations/mysql/wire/util.go [hunk 1]
```diff
@@ -12,6 +13,88 @@ import (
...
+}
+
+// PreparedStmtHistory maintains full history of prepared statements
+// to handle ID reuse scenarios (prepare→close→prepare cycles)
+
+type PreparedStmtHistory struct {
+	Entries      []PreparedStmtEntry // Append-only history of all preparations
+	QueryIndex   map[string][]int    // normalized query -> indices in Entries
+	CurrentCycle uint32              // Monotonically increasing prepare counter
+}
```
[the remark this answers is not in the document]
Comment (codeMaestro78) [reply pkg/agent/proxy/integrations/mysql/wire/util.go:35]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/keploy/keploy/pull/3337#discussion_r2592313963)

Good observation! The PreparedStmtHistory struct is intentionally designed without a mutex because DecodeContext (which contains [StmtHistory](vscode-file://vscode-app/c:/Users/Devarshi/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html)) is used within a single goroutine per MySQL connection during replay. Each connection gets its own DecodeContext instance initialized in [replay.go](vscode-file://vscode-app/c:/Users/Devarshi/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html).

If concurrent access becomes necessary in the future, we can add sync.RWMutex similar to how LastOperation and [ServerGreetings](vscode-file://vscode-app/c:/Users/Devarshi/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) are implemented in the same file ([util.go](vscode-file://vscode-app/c:/Users/Devarshi/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html)). For now, keeping it mutex-free avoids unnecessary overhead for the single-goroutine use case.

File: pkg/agent/proxy/integrations/mysql/replayer/match.go [hunk 2]
```diff
@@ -22,10 +22,12 @@ import (
...
-type prepEntry struct { // minimal, enough for lookup
-	statementID uint32
-	query       string
-	mockName    string // for debugging purpose
+type prepEntry struct {
+	statementID  uint32
+	query        string
+	mockName     string // for debugging purpose
+	prepareOrder int    // Order of preparation for this query (1st, 2nd, 3rd time prepared)
+	wasClosed    bool   // Whether this statement was closed before another prepare
```
[the remark this answers is not in the document]
Comment (codeMaestro78) [reply pkg/agent/proxy/integrations/mysql/replayer/match.go:38]: The [prepareOrder](vscode-file://vscode-app/c:/Users/Devarshi/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) and [wasClosed](vscode-file://vscode-app/c:/Users/Devarshi/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) fields serve two purposes:

Current use: [markClosedEntries()](vscode-file://vscode-app/c:/Users/Devarshi/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) uses [wasClosed](vscode-file://vscode-app/c:/Users/Devarshi/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) to track which recorded statements were closed, enabling accurate lifecycle tracking.

Matching works via query comparison: The current matching logic in [matchStmtExecutePacketQueryAware](vscode-file://vscode-app/c:/Users/Devarshi/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) compares query content (via [expectedQuery](vscode-file://vscode-app/c:/Users/Devarshi/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) and [actualQuery](vscode-file://vscode-app/c:/Users/Devarshi/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html)) rather than statement IDs. This handles ID reuse cases because:

If recorded mock has [stmtID=1](vscode-file://vscode-app/c:/Users/Devarshi/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) with query "SELECT * FROM users"
And runtime has [stmtID=100](vscode-file://vscode-app/c:/Users/Devarshi/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) with the same query
The match succeeds because queries match, not IDs
Future enhancement opportunity: The [prepareOrder](vscode-file://vscode-app/c:/Users/Devarshi/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) field enables future work to correlate by prepare order when you have multiple different queries interleaved in complex patterns. I've added documentation explaining this.
